[DDP] Refactor and remove sync_params

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64515
* __->__ #64514
* #64513

sync_params is a misnomer since we don't actually synchroniz
parameters. While removing this I realized
`self._check_and_sync_module_buffers` does almost everything we need it to, so
just refactored that and made DDP forward call into it.

Differential Revision: [D30751231](https://our.internmc.facebook.com/intern/diff/D30751231/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

rohan-varma committed:
[DDP] Refactor and remove sync_params

sync_params is a misnomer since we don't actually synchroniz
parameters. While removing this I realized
`self._check_and_sync_module_buffers` does almost everything we need it to, so
just refactored that and made DDP forward call into it.

Differential Revision: [D30751231](https://our.internmc.facebook.com/intern/diff/D30751231/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Refactor and remove sync_params"


sync_params is a misnomer since we don't actually synchroniz
parameters. While removing this I realized
`self._check_and_sync_module_buffers` does almost everything we need it to, so
just refactored that and made DDP forward call into it.

Differential Revision: [D30751231](https://our.internmc.facebook.com/intern/diff/D30751231/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -895,7 +895,7 @@ def forward(self, *inputs, **kwargs):
             # specified as part of hook, if hook was specified.
             buffer_hook_registered = hasattr(self, 'buffer_hook')
             if self.require_forward_param_sync and (not buffer_hook_registered or self.buffer_hook.buffer_comm_hook_location == _BufferCommHookLocation.PRE_FORWARD):
```
nit: put this into a helper function, it could also include 'if self.will_sync_module_buffers()' check in this helper function

zhaojuanmao replied:
e.g., will_sync_module_buffers() can include require_forward_param_sync check; should_sync_buffer_forward(), should_sync_buffer_backward()

rohan-varma committed:
Update on "[DDP] Refactor and remove sync_params"


sync_params is a misnomer since we don't actually synchroniz
parameters. While removing this I realized
`self._check_and_sync_module_buffers` does almost everything we need it to, so
just refactored that and made DDP forward call into it.

Differential Revision: [D30751231](https://our.internmc.facebook.com/intern/diff/D30751231/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Refactor and remove sync_params"


sync_params is a misnomer since we don't actually synchroniz
parameters. While removing this I realized
`self._check_and_sync_module_buffers` does almost everything we need it to, so
just refactored that and made DDP forward call into it.

Differential Revision: [D30751231](https://our.internmc.facebook.com/intern/diff/D30751231/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Refactor and remove sync_params"


sync_params is a misnomer since we don't actually synchroniz
parameters. While removing this I realized
`self._check_and_sync_module_buffers` does almost everything we need it to, so
just refactored that and made DDP forward call into it.

Differential Revision: [D30751231](https://our.internmc.facebook.com/intern/diff/D30751231/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Refactor and remove sync_params"


sync_params is a misnomer since we don't actually synchroniz
parameters. While removing this I realized
`self._check_and_sync_module_buffers` does almost everything we need it to, so
just refactored that and made DDP forward call into it.

Differential Revision: [D30751231](https://our.internmc.facebook.com/intern/diff/D30751231/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Refactor and remove sync_params"


sync_params is a misnomer since we don't actually synchroniz
parameters. While removing this I realized
`self._check_and_sync_module_buffers` does almost everything we need it to, so
just refactored that and made DDP forward call into it.

Differential Revision: [D30751231](https://our.internmc.facebook.com/intern/diff/D30751231/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Refactor and remove sync_params"


sync_params is a misnomer since we don't actually synchroniz
parameters. While removing this I realized
`self._check_and_sync_module_buffers` does almost everything we need it to, so
just refactored that and made DDP forward call into it.

Differential Revision: [D30751231](https://our.internmc.facebook.com/intern/diff/D30751231/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]
